### ISSUE TITLE
[HOLD on #26490] [$500] Chat - Space below bold text increases after edit

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#29704]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.84.7
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1697477034366019

## Action Performed:
1. Open the app
2. Open any report
3. Send bold text with # and send code block with 3 backtick in one line eg: # test ```Hii```
4. Edit the message and modify text inside code block and save
5. Observe that now space between bold text and code block increases

## Expected Result:
Describe what you think should've happened

## Actual Result:
Space below bold text increases on edit even when we don't add more space during edit if the text below bold text is code block

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

 

https://github.com/Expensify/App/assets/93399543/c4cfb5b1-0d56-4e82-a620-6e2f43c2fd01

</details>

<details>
<summary>Android: mWeb Chrome</summary>

 

https://github.com/Expensify/App/assets/93399543/57463222-5feb-4cd0-a6f6-6754c2aa6a6e

</details>

<details>
<summary>iOS: Native</summary>

  

https://github.com/Expensify/App/assets/93399543/1347dd28-fb87-4095-841f-2084af2f22c8

</details>

<details>
<summary>iOS: mWeb Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/16154a9b-8f45-444a-8a79-2c94dec725b0

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/93399543/7aca40a5-ce45-4c20-bcca-ed8cb2c0a84f

https://github.com/Expensify/App/assets/93399543/626461f6-5e6f-4dae-a99b-f2620846103c

https://github.com/Expensify/App/assets/93399543/d35b0ffb-37df-4081-8894-ddf4497f81e1

</details>

<details>
<summary>MacOS: Desktop</summary>

 

https://github.com/Expensify/App/assets/93399543/d50f794b-896a-43a6-b04c-d3a0a79bdea4

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b06817cf280a1ec0</li>
        <li>Upwork Job ID: 1713990028027006976</li>
        <li>Last Price Increase: 2023-11-06</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @twisterdotcom</details>

**Comment by tjbo** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Unintended spaces are added after a user edits a markdown message.

### What is the root cause of that problem?

There are 3 main call sites on the component side where this field gets updated:

1. in MessageComposer when the message is first submitted
2. UpdateDraft() - when editing the message on every key stroke in [ReportActionItemMessageEdit.js#L264](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L264)
3. PublishDraft() - when we finish editing the markdown message also in [ReportActionItemMessageEdit.js#L317](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L317)

If we disable #3 in this list everything works fine, except that #3 does a bunch of string processing for many different reasons.

This PublishDraft() function calls something called `Report.editReportComment()` which calls `handleUserDeletedLinksInHtml()`.

[This is where the issue occurs](https://github.com/Expensify/App/blob/main/src/libs/actions/Report.js#L1174).

The HTML is not completely cleaned, there are is hidden spaces between the tags that later become line breaks when coverted to Markdown and then get passed over later into HTML on subsequent publishes.

It doesn't happen the first time you compose the message because that message is done with a different editor (#1 above).

### What changes do you think we should make in order to solve the problem?

[Modify the 1174 of Report.js with:](https://github.com/Expensify/App/blob/main/src/libs/actions/Report.js#L1174).

```const htmlForNewComment = parser.replace(newCommentText).replace(/>\s+</g, '><')```

This will clean the spaces between the HTML tags.

An alternative fix might be to add a rule to the ExpensiMarker lib, that "cleans HTML" after parsing and conversion.

### What alternative solutions did you explore? (Optional)

I explored all call sites on the components side where the text was parsed, saved or rendered.

**Comment by tjbo** [comment]
Contributor details
Your Expensify account email: tom@prototypable.ca
Upwork Profile Link: https://www.upwork.com/freelancers/tomoc

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Chat - Space below bold text increases after edit.

### What is the root cause of that problem?

The root cause of this issue is that when adding a comment or saving draft comment below

````
# heading
```
code
```
````

we call ExpensiMark.replace method to first apply [codeFence](https://github.com/Expensify/expensify-common/blob/6517be47d31aef07b15c90fdf347a341f3b9b58d/lib/ExpensiMark.js#L24-L38) rule, then [heading](https://github.com/Expensify/expensify-common/blob/6517be47d31aef07b15c90fdf347a341f3b9b58d/lib/ExpensiMark.js#L86-L90) rule and last [newline](https://github.com/Expensify/expensify-common/blob/6517be47d31aef07b15c90fdf347a341f3b9b58d/lib/ExpensiMark.js#L221-L225) rule. After applying `codeFence` and `heading` rule, we get intermediate result

```
<h1>heading</h1>
<pre>code test<br /></pre>
```

The line break after `</h1>` results in the unexpected space between heading and code block.

### What changes do you think we should make in order to solve the problem?

When translating markdown text into HTML, the single line break between two block tags should be removed. Currently, we only support 3 block HTML tags, `(blockquote|h1|pre)`.

We can add a new rule before the  [newline](https://github.com/Expensify/expensify-common/blob/6517be47d31aef07b15c90fdf347a341f3b9b58d/lib/ExpensiMark.js#L221-L225) rule to eliminate the unwanted line break between two block elements, like

```js
            {
                name: 'removeNewlineBetweenTwoBlockElements',
                regex: /<\/(blockquote|h1|pre)>\r?\n<(blockquote|h1|pre)>/g,
                replacement: `</$1><$2>`,
            },
```

By having the new rule, we can fix the issue and pass all existing tests of Expensify-common lib.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by eh2077** [comment]
> Might have same root cause as #26490

@0xmiroslav They're related but their root cause is different. This issue occurs when adding a comment and the root cause is from the markdown-to-html process.

**Comment by nikosamofa** [comment]
@0xmiroslav I think you're right, they are related.
The proposal https://github.com/Expensify/App/issues/26490#issuecomment-1758233329 will resolve the root issue.
The recording is based on the proposal.

https://github.com/Expensify/App/assets/30532000/e7ed2885-ebac-42ed-b0b7-beedb63d9783

**Comment by twisterdotcom** [comment]
Nice, okay, shall we close this? And I'll pay out the bug reporter: @dhanashree-sawant?

**Comment by nikosamofa** [comment]
@twisterdotcom I guess not yet. This is not on production yet.
@0xmiroslav will return with his feedback for the best approach on Monday and then we can see.

**Comment by eh2077** [comment]
@twisterdotcom While I think their root cause is different and they should be fixed individually. I think we can wait C+ team to evaluate it.

cc @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
Will review today.

**Comment by akamefi202** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Space below bold text increases after edit.

### What is the root cause of that problem?
```
{
    // We're removing <br /> because when </pre> and <br /> occur together, an extra line is added.
    name: 'replacepre',
    regex: /<\/pre>\s*<br\s*[/]?>/gi,
    replacement: '</pre>',
},
{
    // We're removing <br /> because when <br /> and <pre> occur together, an extra line is added.
    name: 'replacebr',
    regex: /<br\s*[/]?><pre>\s*/gi,
    replacement: ' <pre>',
},
{
    // We're removing <br /> because when <h1> and <br /> occur together, an extra line is added.
    name: 'replaceh1br',
    regex: /<\/h1><br\s*[/]?>/gi,
    replacement: '</h1>',
},
```
The root cause is these 3 replace regex rules in `ExpensiMark` class of `expensify-common` repo.
`replacebr` rule replaces the linebreak with a space if the linkbreak exists before `<pre>` tag.
And it renders the linebreak after the space while rendering in the report page.

This doesn't cause problem if previous line is a normal text.
But if the previous line is `</h1>` tag, we see an extra linebreak because `</h1>` tag renders a linebreak after it too.

For example, like below.
````
# test\n```code``` => <h1>test</h1>\n<pre>code</pre> => <h1>test</h1> <pre>code</pre> => TEST\n \nCODE
````

### What changes do you think we should make in order to solve the problem?
```
{
    // We're removing <br /> because when <h1> and <br /> occur together, an extra line is added.
    name: 'replaceh1br',
    regex: /<\/h1><br\s*[/]?>/gi,
    replacement: '</h1>',
},
{
    // We're removing <br /> because when </pre> and <br /> occur together, an extra line is added.
    name: 'replacepre',
    regex: /<\/pre>\s*<br\s*[/]?>/gi,
    replacement: '</pre>',
},
{
    // We're removing <br /> because when <br /> and <pre> occur together, an extra line is added.
    name: 'replacebr',
    regex: /<br\s*[/]?><pre>\s*/gi,
    replacement: ' <pre>',
},
```
To fix this issue, we need to prioritize `replaceh1br` rule before other two rules.
So `replacebr` rule won't add unnecessary space and we won't see an extra line too.

For example, like below.
````
# test\n```code``` => <h1>test</h1>\n<pre>code</pre> => <h1>test</h1><pre>code</pre> => TEST\nCODE
````

### What alternative solutions did you explore? (Optional)
N/A

**Comment by abdulrahuman5196** [comment]
@eh2077 
>We can add a new rule before the [newline](https://github.com/Expensify/expensify-common/blob/6517be47d31aef07b15c90fdf347a341f3b9b58d/lib/ExpensiMark.js#L221-L225) rule to eliminate the unwanted line break between two block elements

Why should we do this? What if the user actually wanted to add a new line inbetween?

**Comment by eh2077** [comment]
> @eh2077
> 
> > We can add a new rule before the [newline](https://github.com/Expensify/expensify-common/blob/6517be47d31aef07b15c90fdf347a341f3b9b58d/lib/ExpensiMark.js#L221-L225) rule to eliminate the unwanted line break between two block elements
> 
> Why should we do this? What if the user actually wanted to add a new line inbetween?

@abdulrahuman5196 Good point. There'll be two line breaks if we want to add a new line between two block elements. For example, the following input

````
# heading

```code```
````
will be translated into intermediate result, `<h1>heading</h1>\n\n<pre>code</pre>`, which will be skipped by the replacement rule I added. The new rule 
```js
            {
                name: 'removeNewlineBetweenTwoBlockElements',
                regex: /<\/(blockquote|h1|pre)>\r?\n<(blockquote|h1|pre)>/g,
                replacement: `</$1><$2>`,
            },
```
will only handle case with one line break between two block tags.

**Comment by akamefi202** [comment]
@abdulrahuman5196 Could you please let me know your opinion about my proposal too?
https://github.com/Expensify/App/issues/29704#issuecomment-1777484371

**Comment by abdulrahuman5196** [comment]
Will work on pending reviews today.

**Comment by nikosamofa** [comment]
@abdulrahuman5196 Can you also review https://github.com/Expensify/App/issues/26490#issuecomment-1758233329 ? I think they are related and the proposal can resolve 29704 too

**Comment by abdulrahuman5196** [comment]
Sorry for the delay. Will review today

**Comment by abdulrahuman5196** [comment]
TAL

**Comment by twisterdotcom** [comment]
What does TAL mean?

**Comment by abdulrahuman5196** [comment]
Hi, Sorr,y TAL means Taking a look.
Will review again today, just back from weekend

**Comment by twisterdotcom** [comment]
@abdulrahuman5196 are you able to get somebody assigned here or can I reassign to a new C+? Will reassign on Monday if you're not around by then.

**Comment by twisterdotcom** [comment]
Not overdue, just waiting on @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
Sure @twisterdotcom . I will cover this over the weekend. If not do go ahead and re-assign on monday.

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 Can you also review [#26490 (comment)](https://github.com/Expensify/App/issues/26490#issuecomment-1758233329) ? I think they are related and the proposal can resolve 29704 too

With this change, I am unable to reproduce the issue. The proposal is already approved as part of https://github.com/Expensify/App/issues/26490.

Can we put this issue on hold for https://github.com/Expensify/App/issues/26490? @twisterdotcom 



https://github.com/Expensify/App/assets/46707890/c4ebe04e-2ab7-4e65-8b7a-b0732da0cfe7

**Comment by twisterdotcom** [comment]
Done

**Comment by abdulrahuman5196** [comment]
Let me check in my morning
